kong/tools/wrpc/future.lua

```diff
@@ -0,0 +1,123 @@
+local semaphore = require "ngx.semaphore"
+local semaphore_new = semaphore.new
+
+local ngx_log = ngx.log
+local ERR = ngx.ERR
+local ngx_now = ngx.now
+local new_timer = ngx.timer.at
+local setmetatable = setmetatable
+
+-- This type works like a JavaScript promise.
+-- You can call `then_do` to have haviour similar to JS `then`
+-- and call `wait` for async call.
+---@class kong.tools.wrpc.future
+local _M = {}
+local _MT = { __index = _M, }
+
+local function finish_future_life(future)
+  future.responses_tab[future.seq] = nil
+end
+
+local function drop_aftermath(premature, future)
+  if premature then
+    return
+  end
+
+  local ok, err = future:wait()
+  if not ok then
+    ngx_log(ERR, "request fail to recieve response: ", err)
+  end
+end
+
+-- Call to intentionally drop the future, without blocking to wait.
+-- It will discard the response and log if error occurs.
+--- @return boolean ok, string err
+function _M:drop()
+  return new_timer(0, drop_aftermath, self)
+end
+
+local function then_do_handle_result(premature, future, hdl, err_hdl)
```

Comment:
This PR is not about those parts. It's just based on the previous PR. Maybe after that PR passed I will rebase it.

Reply:
you can retarget this PR  against `refactor/wrpc` instead of `master`, so the diff would show only this changes.  when #8883 is merged, this PR is automatically retargeted to `master`